system/logaccess/ElasticStack/ElasticStackLogAccess.cpp

```diff
@@ -0,0 +1,502 @@
+    }
+    catch (...)
+    {
+        LOG(MCuserError, "ESLogAccess: Encountered error requesting '%s'", logmessageprefix);
+    }
+    return nullptr;
+
+}
+
+const IPropertyTree * ElasticStackLogAccess::getTimestampTypeFormat(const char * indexpattern, const char * fieldname)
+{
+    if (isEmptyString(indexpattern))
+        throw makeStringException(-1, "ElasticStackLogAccess::getTimestampTypeFormat: indexpattern must be provided");
+
+    if (isEmptyString(fieldname))
+        throw makeStringException(-1, "ElasticStackLogAccess::getTimestampTypeFormat: fieldname must be provided");
+
+    VStringBuffer timestampformatreq("%s/_mapping/field/created_ts?include_type_name=true&format=JSON", indexpattern);
+    return performAndLogESRequest(Client::HTTPMethod::GET, timestampformatreq.str(), "", "getTimestampTypeFormat");
+}
+
+const IPropertyTree * ElasticStackLogAccess::getIndexSearchStatus(const char * indexpattern)
+{
+    if (!indexpattern || !*indexpattern)
+        throw makeStringException(-1, "ElasticStackLogAccess::getIndexSearchStatus: indexpattern must be provided");
+
+    VStringBuffer indexsearch("_cat/indices/%s?format=JSON", indexpattern);
+    return performAndLogESRequest(Client::HTTPMethod::GET, indexsearch.str(), "", "List of available indexes");
+
+}
+
+const IPropertyTree * ElasticStackLogAccess::getESStatus()
+{
+    return performAndLogESRequest(Client::HTTPMethod::GET, "_cluster/health", "", "Target cluster health");
+}
+
+/*
+ * Transform ES query response to back-end agnostic response
+ *
+ */
+void processESJsonResp(const cpr::Response & retrievedDocument, StringBuffer & returnbuf, LogAccessLogFormat format)
+{
+    if (retrievedDocument.status_code != 200)
+        throw makeStringExceptionV(-1, "ElasticSearch request failed: %s", retrievedDocument.status_line.c_str());
+
+    DBGLOG("Retrieved ES JSON DOC: %s", retrievedDocument.text.c_str());
+
+    Owned<IPropertyTree> tree = createPTreeFromJSONString(retrievedDocument.text.c_str());
+    if (!tree)
+        throw makeStringExceptionV(-1, "%s: Could not parse ElasticSearch query response", COMPONENT_NAME);
+
+    if (tree->hasProp("timed_out") && tree->getPropBool("timed_out"))
+        LOG(MCuserProgress,"ES Log Access: timeout reported");
+    if (tree->hasProp("_shards/failed") && tree->getPropInt("_shards/failed") > 0)
+        LOG(MCuserProgress,"ES Log Access: failed _shards reported");
+
+    DBGLOG("ES Log Access: hit count: '%d'", tree->getPropInt("hits/total/value"));
+
+    Owned<IPropertyTreeIterator> iter = tree->getElements("hits/hits/fields");
+    switch (format)
+    {
+        case LOGACCESS_LOGFORMAT_xml:
+        {
+            returnbuf.set("<lines>");
+            StringBuffer hitchildxml;
+
+            ForEach(*iter)
+            {
+                IPropertyTree & cur = iter->query();
+                toXML(&cur,hitchildxml.clear());
```

Comment:
note: related to comment above, this could be coded as
```
returnbuf.append("<line>");
toXML(&cur,returnbug);
returnbuf.append("</line>");
```
avoiding the temporary variable.

Reply:
but the trade-off is an extra call to the StringBuffer's append per iteration... can this cause extra resizes?